### PR TITLE
Fix labels perma diff

### PR DESCRIPTION
**Pull request by iwahbe**
This PR takes advantage of https://github.com/pulumi/pulumi-terraform-bridge/pull/2417 to fixup the incorrect planned state caused by empty labels.

Fixes https://github.com/pulumi/pulumi-gcp/issues/2372

--- 

Before merging
- [x] tests need to be added

### CODE REVIEW

File: provider/resources.go [hunk 1]
```diff
@@ -454,14 +455,72 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
+// https://github.com/pulumi/pulumi-gcp/issues/2372.
+func fixEmptyLabels(_ context.Context, req shimv2.PlanStateEdit) (cty.Value, error) {
+	inputs, ok := resource.PropertyPath{"labels"}.Get(resource.NewProperty(req.NewInputs))
+	if !ok {
+		return req.PlanState, nil
+	}
+
+	labels := resource.FromResourcePropertyValue(inputs)
+	if !labels.IsMap() {
+		// No labels found, so we don't need to correct
```
Comment (VenelinMartinov) [inline provider/resources.go:471]: I don't thin that's true - the resource could have `effective_labels` or `terraform_labels` without having any `labels` set.

File: provider/resources.go [hunk 2]
```diff
@@ -454,14 +455,72 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
+		return req.PlanState, nil
+	}
+
+	// fixOutputLabels fixes falsely computed values inherited from "labels".
```
Comment (VenelinMartinov) [inline provider/resources.go:475]: Do you mean "unknown" not "computed"? I believe this are all computed either way.

File: provider/resources.go [hunk 3]
```diff
@@ -454,14 +455,72 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
+		m := output.AsValueMap()
+		for k, v := range m {
+			if v.IsKnown() {
+				// If v is known, then it is not falsely computed. No
+				// action is needed.
+				continue
+			}
+			label, ok := labels.AsMap()[k] // labels is in the Pulumi namespace
+			if !ok || label.IsComputed() || !label.IsString() {
+				// If we didn't inherit label from "labels" or the label
```
Comment (VenelinMartinov) [inline provider/resources.go:489]: Does the issue not happen with empty labels in defaultLabels/effectiveLabels?

Comment (iwahbe) [reply provider/resources.go:499]: It did, and that was addressed.

File: provider/resources.go (hunk 2 above)
Comment (iwahbe) [reply provider/resources.go:485]: I meant computed from the Pulumi namespace, which is equivalent to the TF concept of unknown.

**Review by VenelinMartinov — COMMENTED** [review]
Can we also add a test for unmanaged labels? You can use the gcloud CLI:

```
gcloud storage buckets update $bucket_name --update-labels=unmanaged=value,unmanaged_empty=
```

File: provider/resources.go [hunk 4]
```diff
@@ -454,14 +456,82 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
 //go:embed cmd/pulumi-resource-gcp/bridge-metadata.json
 var metadata []byte
 
+// A predicate function that always returns true.
+func always[T any](T) bool { return true }
+
+// fixEmptyLabels applies a state edit to fix
+// https://github.com/pulumi/pulumi-gcp/issues/2372.
+func fixEmptyLabels(_ context.Context, req shimv2.PlanStateEditRequest) (cty.Value, error) {
```
Comment (VenelinMartinov) [inline provider/resources.go:464]: Should we also add a few unit tests on this to better illustrate what we are dealing with and what we are changing? Might be useful for the next person who has to look into this.

File: provider/resources.go [hunk 5]
```diff
@@ -454,14 +456,82 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
+	// actualLabels holds the labels that we expect to be applied. It is the union of
+	// labels (from the resource) and defaultLabels (from the provider)
```
Comment (VenelinMartinov) [inline provider/resources.go:466]: This does not look correct - the resource might have unmanaged labels which we'd clobber if we apply this.

**Comment by VenelinMartinov** [comment]
I don't think we are handling unmanaged labels correctly here. We should add a test for that case, both for empty and non-empty unmanaged labels.

File: provider/resources.go (hunk 5 above)
Comment (iwahbe) [reply provider/resources.go:466]: This is more a naming issue then a semantic issue. `actualLabels` holds the labels that we expect to be applied. It is not the set of labels that we will apply. It is used only to manage unknowns.

**Comment by iwahbe** [comment]
> I don't think we are handling unmanaged labels correctly here. We should add a test for that case, both for empty and non-empty unmanaged labels.

It doesn't clober, though the upstream refresh behavior isn't great. I'll add a test and see if I can improve upstream's behavior.

**Comment by iwahbe** [comment]
I added a regression test, but I'm not going to tackle fixing labels in this PR. I opened https://github.com/pulumi/pulumi-gcp/issues/2390 to track general unmanaged labels work.

**Comment by VenelinMartinov** [comment]
Can you please clarify what the actual problem is in https://github.com/pulumi/pulumi-gcp/issues/2390? Is it the effectiveLabels output changing during Refresh? If so, that seems expected.

**Comment by iwahbe** [comment]
> Can you please clarify what the actual problem is in #2390? Is it the effectiveLabels output changing during Refresh? If so, that seems expected.

No, it's that it changes during the subsequent update, not during the refresh. I have clarified in the issue.

File: provider/resources.go [hunk 6]
```diff
@@ -454,14 +456,94 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
...
+		for k, v := range m {
+			if v.IsKnown() {
+				// If v is known, so no correction is needed.
+				continue
+			}
+			label, ok := programLabels[k] // labels is in the Pulumi namespace
+			if !ok || label.IsComputed() || !label.IsString() {
+				// If we didn't inherit label from the program or the
+				// label is actually unknown, then just continue.
+				continue
```
Comment (VenelinMartinov) [inline provider/resources.go:514]: We are skipping unmanaged labels here - why does `effectiveLabels` return the correct result for them?

Comment (iwahbe) [reply provider/resources.go:514]: We are choosing not to correct unmanaged labels here. This isn't dropping the skipped labels.

Comment (VenelinMartinov) [reply provider/resources.go:514]: >We are choosing not to correct unmanaged labels here

Wouldn't this mean that an empty unmanaged label causes a diff like before?

**Review by VenelinMartinov — COMMENTED** [review]
Looks good - glad this worked out!

Two points here:

1. Unit tests on the fixup function would be nice for whoever has to interact with this next - it helps to illustrate the desired behaviour as it is not obvious what the GCP provider returns when we interact with the Plan - happy to punt this for after fixing the issue if we want to ship sooner.
2. Why do we not fix unmanaged labels? Why do we treat them differently?

File: provider/resources.go (hunk 6 above)
Comment (guineveresaenger) [reply provider/resources.go:514]: Can we agree on a precise definition for "unmanaged" labels and keep that discoverable somewhere? Do we mean "unmanaged by Pulumi" "unmanaged by Terraform" or something else?

File: provider/provider_yaml_test.go [hunk 7]
```diff
@@ -1120,3 +1124,215 @@ func TestFirestoreDatabaseAutoname(t *testing.T) {
 	pt.SetConfig("gcpProj", proj)
 	pt.Up()
 }
+
+func TestEmptyLabels(t *testing.T) {
```
Comment (guineveresaenger) [inline provider/provider_yaml_test.go:1128]: How do we expect empty labels to behave? Can that be part of the test name?
a silly example - `TestEmptyLabelsDefaultToCake` - tells me Pulumi should rewrite all empty labels to say "cake".

File: provider/resources.go [hunk 8]
```diff
@@ -454,14 +456,94 @@ func preConfigureCallbackWithLogger(credentialsValidationRun *atomic.Bool, gcpCl
 //go:embed cmd/pulumi-resource-gcp/bridge-metadata.json
 var metadata []byte
 
+// A predicate function that always returns true.
```
Comment (guineveresaenger) [inline provider/resources.go:459]: Should all of this code live in `labels.go`?

**Review by guineveresaenger — COMMENTED** [review]
Thank you for linking the outstanding work in the bridge and follow-up.

**Comment by guineveresaenger** [comment]
Also - this needs https://github.com/pulumi/pulumi-terraform-bridge/pull/2417 to merge and even release before we can merge here, right?

**Comment by VenelinMartinov** [comment]
The issue with `TestImport` seems to be an engine regression, opened https://github.com/pulumi/pulumi/issues/17290

File: provider/resources.go (hunk 6 above)
Comment (iwahbe) [reply provider/resources.go:514]: >> We are choosing not to correct unmanaged labels here
> 
> Wouldn't this mean that an empty unmanaged label causes a diff like before?

Observationally they don't. I think that's because the diff customizer doesn't introduce any changes to labels imported via `pulumi refresh`.

> Can we agree on a precise definition for "unmanaged" labels and keep that discoverable somewhere?

I'm not sure where we would keep the definition. Our working definition is "unmanaged" means not managed by IaC (either TF or Pulumi).

File: provider/provider_yaml_test.go [hunk 9]
```diff
@@ -1120,3 +1124,215 @@ func TestFirestoreDatabaseAutoname(t *testing.T) {
...
+// doesn't get worse.
+//
+// We would prefer the obvious behavior patter:
+//
+// $ pulumi up --yes
+// $ pulumi refresh --yes # Retrieves labels from GCP
+// $ pulumi preview --expect-no-changes
+//
+// Improvement tracked in <https://github.com/pulumi/pulumi-gcp/issues/2390>.
+func TestUnmanagedEmptyLabels(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply provider/provider_yaml_test.go:1285]: The comment on `TestUnmanagedEmptyLabels` explains what behavior it is asserting and why. Right now, it's just to ensure that the behavior doesn't change.
